### ISSUE TITLE
Write React / JS / Node / API Tests

### ISSUE THREAD

**Issue by bmorelli25** [bmorelli25/ReactWeather#14]
Hello [Hacktoberfest](https://hacktoberfest.digitalocean.com/)!

Are you interested in React JS, Node JS, APIs, HTML, Javascript, Webpack, Express, and/or SASS/CSS? Then you've come to the right project!

This web app allows a user to type in a location, then pulls weather data from [OpenWeatherMap's API](http://openweathermap.org/) and returns it to the user. **[Live Demo](http://still-springs-11020.herokuapp.com/)**

Your task: Testing! Practice your testing skills by writing tests for this web app. There are a number of things that can be tested in this app, including:
- API data fetch
- API KEY
- React Rendering
- Search in top right box
- Search in main box
- Search with blank string / bad characters

Don't hesitate to reach out with questions, or if you need help understanding the code!

**If you want to write a test(s), leave a comment specifying what tests you're going to write so others don't end up writing the same tests. Tests should be placed in the `/app/tests/` folder. Please see PR #7 for more information on how tests are implemented in this project**

If you're not sure where to start, check out my [React Todo App](https://github.com/bmorelli25/ReactTodo). This Web app has a ton (30+) tests and should give you some great ideas for how to write tests. 

_PS: This project utilizes a `.editorconfig` file. This helps developers define and maintain consistent coding styles between different editors and IDEs. Please ensure your editor is set up to use the `.editorconfig` file. For more information, check out [editorconfig.org](http://editorconfig.org/)._

**Comment by tetreault** [comment]
I'd love to get some more testing experience with mocha/chai - mind if I take tests for **API data fetching**? What sort of testing might need to be done? just expected/unexpected results from OpenWeatherMap? I was going to try basing them off of prior work you did [here](https://github.com/bmorelli25/ReactTodo/blob/master/app/tests/api/TodoAPI.test.jsx). 

Cool project, thanks for making it accessible!

**Comment by bmorelli25** [comment]
Hey @tetreault! That would be great!

I just merged #15 so the app now utilizes two API's (both from OpenWeatherMap). Each one can be tested - although the tests will almost be exactly the same for each.

You can test either the API request itself (e.g. writing a test that actually fetches data from the API, and ensuring the data is returned in the correct format). Or, you can do something like what I did in the link you referenced: Create a variable with mock JSON data that matches what a successful API request would return. Then test ensure the app handles that JSON data correctly. 

Don't forget, an important part of testing is also passing in bad data, or making bad API requests. Ensure the app correctly throws an error and the user is well informed of what went wrong.

Does that help at all? Let me know if you have more questions and thanks again for helping out!

**Comment by tetreault** [comment]
Yeah that's great help! I'm getting things set up locally and will be able to start getting to it later tonight/tomorrow AM! Thanks again!

**Comment by tetreault** [comment]
BTW - is it okay if I use ES6 const & let in the tests instead of var (as I noticed was already being done) or would you prefer just sticking to var?

**Comment by bmorelli25** [comment]
@tetreault Heck yes - I would prefer if you update to ES6

**Comment by tetreault** [comment]
Sorry for the probably naive question, but in testing the API I'm doing something like the following: 

`
const OpenWeatherMap = require('../../api/openWeatherMap.jsx');
it('OpenWeatherMap should fetch weather for a given location', () => {
	let fetchWeatherPromise = OpenWeatherMap.getCurrentWeather('New York City');
});
`

`fetchWeatherPromise` yields `Promise { <pending> }` and normally I thought `.done()` is available yet that's undefined. Any guiding tips on what I should be doing to peek into OpenWeatherMap? I see in the openWeatherMap.jsx file, if successful it returns `res.data;`

Thanks!

**Comment by piq9117** [comment]
Hi! I code in http://standardjs.com/ style is that okay? Basically 2 spaces/1 tab. Single quotation marks, no semi-colons.

Or if you have an eslint rule, i'll gladly follow it. 😄

**Comment by tetreault** [comment]
Hi @piq9117 there's a section in the readme about the style guide

**Comment by piq9117** [comment]
@tetreault did you mean the [foundation](http://foundation.zurb.com/) styling framework? I was talking about about [standardjs](http://standardjs.com/) coding style

**Comment by tetreault** [comment]
Nope, sorry actually I was wrong it wasn't in the readme but in their first comment: 

**PS: This project utilizes a .editorconfig file. This helps developers define and maintain consistent coding styles between different editors and IDEs. Please ensure your editor is set up to use the .editorconfig file. For more information, check out editorconfig.org.**

**Comment by tetreault** [comment]
Submitted pull request for [API data fetching](https://github.com/bmorelli25/React-Weather-App/pull/21). @bmorelli25 let me know if there's any issues or concerns!

**Comment by tetreault** [comment]
Since it hasn't been claimed yet by anyone else I'll claim writing tests for **search in top right box** and **search in main box**.

**Comment by bmorelli25** [comment]
Sorry, I was a little busy yesterday and couldn't get online. 

@piq9117 - I want you to be able to contribute, but I think that would make the codebase too confusing for newer developers. At this time I want to keep everything standard and as noted in the .editorconfig file

@tetreault  Awesome, Awesome, and more awesome! I'll get to downloading and testing your tests asap

**Comment by tetreault** [comment]
Sorry for the radio silence on my end with this -- busy start of the week at work

**Comment by bmorelli25** [comment]
No worries @tetreault, there is no rush

**Comment by tetreault** [comment]
Hey @bmorelli25 ! Finally got some free time and have a question for you since you know React better than I. How should I go about testing the search functionality done thru react? I was [going to start trying to base things off of the ToDoSearch testing you did in another repo](https://github.com/bmorelli25/ReactTodo/blob/master/app/tests/components/TodoSearch.test.jsx) but am being stopped in my tracks based off of a require statement you have in Nav.jsx. Any advice/guidance is for sure appreciated!

``` shell
var {Link, IndexLink} = require('react-router');
    ^

SyntaxError: Unexpected token {
    at exports.runInThisContext (vm.js:53:16)
    at Module._compile (module.js:387:25)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Module.require (module.js:367:17)
    at require (internal/module.js:16:19)
    at Object.<anonymous> (/Users/tetreault1/Documents/git/React-Weather-App/app/tests/components/searchFields.spec.js:12:13)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Module.require (module.js:367:17)
    at require (internal/module.js:16:19)
    at /Users/tetreault1/Documents/git/React-Weather-App/node_modules/mocha/lib/mocha.js:220:27
    at Array.forEach (native)
    at Mocha.loadFiles (/Users/tetreault1/Documents/git/React-Weather-App/node_modules/mocha/lib/mocha.js:217:14)
    at Mocha.run (/Users/tetreault1/Documents/git/React-Weather-App/node_modules/mocha/lib/mocha.js:485:10)
    at Object.<anonymous> (/Users/tetreault1/Documents/git/React-Weather-App/node_modules/mocha/bin/_mocha:457:18)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Function.Module.runMain (module.js:447:10)
    at startup (node.js:146:18)
    at node.js:404:3
npm ERR! Test failed.  See above for more details.
```

I haven't even really done anything yet just the following: 

``` javascript
'use strict';

const React = require('react');
const ReactAddons = require('react/addons');  // deprecated apparently, react doesn't like this?
const ReactDOM = require('react-dom');
const TestUtils = React.addons.TestUtils;
const assert = require('chai').assert;
const expect = require('chai').expect;

const Nav = require('../../components/Nav.jsx');

// test the Nav Search
describe('Nav Search Testing', () => {
});
```

**Comment by bmorelli25** [comment]
Hey @tetreault. Sorry for the delay on this one. This is a much more complex test than I initially intended. There are a few things that need to be done.

package.json:
`"react-addons-test-utils": "^0.14.7"`

testFile.jsx:
`var TestUtils = require('react-addons-test-utils');`

When testing this component, the test file has to be a `.jsx` file.
The component needs to `export default` or the component needs to be wrapped in `{}` when imported. [Source](http://stackoverflow.com/questions/33956201/how-to-import-and-export-components-using-react-es6-webpack). This means `Nav.jsx` will need to be refactored. Take a look at how [TodoSearch](https://github.com/bmorelli25/ReactTodo/blob/master/app/components/TodoSearch.jsx) exports a default. 

Even with all this, I still can't get it to work quite right. I'm going to keep looking into tonight when I have a little more time. Sorry!

**Comment by tetreault** [comment]
No problem @bmorelli25 - keep me posted! And thanks again for starting a fun project :)

**Comment by jamesmacfie** [comment]
@bmorelli25 Could I include [enzyme](https://github.com/airbnb/enzyme) as a dev dependency for tests? It makes it a lot easier to write React tests :)

**Comment by jamesmacfie** [comment]
Sorry, ignore that last comment. I see that enzyme is included but just isn't being used

**Comment by bmorelli25** [comment]
@tetreault, I cannot for the life of me get this figured out. @jamesmacfie has opened up #33 and he's made more progress than I, so hopefully this is close to being fixed (finally)

**Comment by bmorelli25** [comment]
James' changes have been merged. Testing should finally work again @tetreault!! Opening up a new issue to fix the Aliases

**Comment by tetreault** [comment]
Sweet! Thanks @bmorelli25 -- so @jamesmacfie did you write tests for the react search components? if not I'm still for sure happy to man that one.

**Comment by jamesmacfie** [comment]
@tetreault I started writing some UI tests last night but have not put in a PR yet (take a look [here](https://github.com/jamesmacfie/React-Weather-App/blob/5ed255bf02be51f52adab57ddea8d35425758bd0/app/tests/components/WeatherMessage.spec.js) if you're interested). All I've started confirming here is that given a certain prop, it's rendered somewhere within the resulting HTML (where, specifically, it's rendered is more of an implementation detail that I don't think needs testing. i.e You shouldn't have to change your tests if you move the location from an `h3` to a `p` or something).

The `<Nav />` component is a bit more complex and, like @bmorelli25 mentioned, could need a bit of a refactor so won't be as simple as those tests I just linked to. 

If you manage to get the setup working and are still having issues then yell out and we're happy to lend a hand
